common/util.go

```diff
@@ -410,3 +411,45 @@ func ReadAtMost(r io.Reader, n int) ([]byte, error) {
 	}
 	return b, err
 }
+
+func enumerateNvidiaDevices() []string {
+	gpu, err := ghw.GPU()
+	if err != nil {
+		glog.Errorf("Error getting GPU info: %v", err)
+		return nil
+	}
+
+	nvidiaCardCount := 0
+	for _, card := range gpu.GraphicsCards {
+		if strings.EqualFold(card.DeviceInfo.Vendor.Name[:6], "nvidia") {
+			nvidiaCardCount += 1
+		}
+	}
+	if nvidiaCardCount == 0 {
+		glog.Warning("No nvidia devices found")
+		return nil
+	}
+
+	devices := []string{}
+
+	for i := 0; i < nvidiaCardCount; i++ {
+		s := strconv.Itoa(i)
+		devices = append(devices, s)
+	}
+
+	return devices
+}
+
+func ParseNvidiaDevicesFlag(nvidia string) []string {
+	var devices []string
+	if nvidia == "all" {
+		devices = enumerateNvidiaDevices()
+		if devices == nil || len(devices) == 0 {
+			glog.Warningf("Unable to query available Nvidia gpus, using default device ID '0'")
+			devices = []string{"0"}
```

Comment:
If no devices are found (`devices == 0`) defaulting to a GPU seems likeit could lead to unexpected behaviour or errors (in case there really is no GPU). 

Perhaps a sensible default would be to default to CPU transcoding and clearly log this message. Alternatively, when a user explicitly provides a -nvidia flag the node should always have access to a GPU, in case `devices == 0` the node should exit with an error. 

WDYT? Might be missing something here.

Reply:
> If no devices are found (devices == 0) defaulting to a GPU seems likeit could lead to unexpected behaviour or errors (in case there really is no GPU)

This was done as a fallback for the case when this newly introduced "GPU detection" fails for a user, but a GPU is still present.

If a GPU is not present at all it would anyway fail gracefully during startup with `testTranscoding=true` by default - we currently also allow a user to manually set `-nvidia=0` on a CPU-only machine, relying on that check.

If this gpu detection feature works reliably on different systems, we can choose to rely on this more strongly in future. We'll still ofc need `testTranscoding` to detect session limit caps. WDYT?

Reply:
I'm in favor of being more explicit with the failure case if no devices are found and clearly logging a message indicating that no devices were found instead of allowing startup to proceed and hit the `-testTranscoder` tests. The `-testTranscoder` tests will fail if it is actually the case that no devices are found, but I don't think the error messages are particularly user friendly (someone feel free to question this). So, if there is an opportunity to fail earlier if we know there are no devices available then I think doing so can be more informative for the user.

I also agree with Nico that defaulting to the first device if there is a problem with GPU detection when the user specifies `-nvidia=all` feels like it would be surprising to a user. From my own POV, I would be surprised if I passed in `-nvidia=all` expecting to use all 8 of my GPUs only to realize later on that only 1 GPU is being used. Additionally, if no devices are detected, but there *are* actually devices present due to some issue with the GPU detection feature, then a user is still able to manually specify device IDs if absolutely necessary.

Reply:
> I would be surprised if I passed in -nvidia=all expecting to use all 8 of my GPUs only to realize later on that only 1 GPU is being used.

Yep that makes sense. I've updated the log line, and also using `Fatalf` in the fixup commit.